File: dbflute-scala-testlib/src/main/java/org/dbflute/scala/testlib/dbflute/cbean/cq/bs/BsRegionCQ.java
package org.dbflute.scala.testlib.dbflute.cbean.cq.bs;

import java.util.Map;

import org.dbflute.scala.testlib.dbflute.cbean.*;
import org.dbflute.scala.testlib.dbflute.cbean.cq.*;
import org.dbflute.scala.testlib.dbflute.cbean.cq.ciq.*;
import org.seasar.dbflute.cbean.*;
import org.seasar.dbflute.cbean.cvalue.ConditionValue;
import org.seasar.dbflute.cbean.sqlclause.SqlClause;
import org.seasar.dbflute.exception.IllegalConditionBeanOperationException;

/**
 * The base condition-query of REGION.
 * @author DBFlute(AutoGenerator)
 */
public class BsRegionCQ extends AbstractBsRegionCQ {

    // ===================================================================================
    //                                                                           Attribute
    //                                                                           =========
    protected RegionCIQ _inlineQuery;

    // ===================================================================================
    //                                                                         Constructor
    //                                                                         ===========
    public BsRegionCQ(ConditionQuery childQuery, SqlClause sqlClause, String aliasName, int nestLevel) {
        super(childQuery, sqlClause, aliasName, nestLevel);
    }

    // ===================================================================================
    //                                                                 InlineView/OrClause
    //                                                                 ===================
    /**
     * Prepare InlineView query. <br />
     * {select ... from ... left outer join (select * from REGION) where FOO = [value] ...}
     * <pre>
     * cb.query().queryMemberStatus().<span style="color: #FD4747">inline()</span>.setFoo...;
     * </pre>
     * @return The condition-query for InlineView query. (NotNull)
     */
    public RegionCIQ inline() {
        if (_inlineQuery == null) { _inlineQuery = xcreateCIQ(); }
        _inlineQuery.xsetOnClause(false); return _inlineQuery;
    }

    protected RegionCIQ xcreateCIQ() {
        RegionCIQ ciq = xnewCIQ();
        ciq.xsetBaseCB(_baseCB);
        return ciq;
    }

    protected RegionCIQ xnewCIQ() {
        return new RegionCIQ(xgetReferrerQuery(), xgetSqlClause(), xgetAliasName(), xgetNestLevel(), this);
    }

    /**
     * Prepare OnClause query. <br />
     * {select ... from ... left outer join REGION on ... and FOO = [value] ...}
     * <pre>
     * cb.query().queryMemberStatus().<span style="color: #FD4747">on()</span>.setFoo...;
     * </pre>
     * @return The condition-query for OnClause query. (NotNull)
     * @throws IllegalConditionBeanOperationException When this condition-query is base query.
     */
    public RegionCIQ on() {
        if (isBaseQuery()) { throw new IllegalConditionBeanOperationException("OnClause for local table is unavailable!"); }
        RegionCIQ inlineQuery = inline(); inlineQuery.xsetOnClause(true); return inlineQuery;
    }

    // ===================================================================================
    //                                                                               Query
    //                                                                               =====

    protected ConditionValue _regionId;
    public ConditionValue getRegionId() {
        if (_regionId == null) { _regionId = nCV(); }
        return _regionId;
    }
    protected ConditionValue getCValueRegionId() { return getRegionId(); }

    protected Map<String, MemberAddressCQ> _regionId_ExistsReferrer_MemberAddressListMap;
    public Map<String, MemberAddressCQ> getRegionId_ExistsReferrer_MemberAddressList() { return _regionId_ExistsReferrer_MemberAddressListMap; }
    public String keepRegionId_ExistsReferrer_MemberAddressList(MemberAddressCQ sq) {
        if (_regionId_ExistsReferrer_MemberAddressListMap == null) { _regionId_ExistsReferrer_MemberAddressListMap = newLinkedHashMapSized(4); }
        String ky = "subQueryMapKey" + (_regionId_ExistsReferrer_MemberAddressListMap.size() + 1);
        _regionId_ExistsReferrer_MemberAddressListMap.put(ky, sq); return "regionId_ExistsReferrer_MemberAddressList." + ky;
    }

    protected Map<String, MemberAddressCQ> _regionId_NotExistsReferrer_MemberAddressListMap;
    public Map<String, MemberAddressCQ> getRegionId_NotExistsReferrer_MemberAddressList() { return _regionId_NotExistsReferrer_MemberAddressListMap; }
    public String keepRegionId_NotExistsReferrer_MemberAddressList(MemberAddressCQ sq) {
        if (_regionId_NotExistsReferrer_MemberAddressListMap == null) { _regionId_NotExistsReferrer_MemberAddressListMap = newLinkedHashMapSized(4); }
        String ky = "subQueryMapKey" + (_regionId_NotExistsReferrer_MemberAddressListMap.size() + 1);
        _regionId_NotExistsReferrer_MemberAddressListMap.put(ky, sq); return "regionId_NotExistsReferrer_MemberAddressList." + ky;
    }

    protected Map<String, MemberAddressCQ> _regionId_SpecifyDerivedReferrer_MemberAddressListMap;
    public Map<String, MemberAddressCQ> getRegionId_SpecifyDerivedReferrer_MemberAddressList() { return _regionId_SpecifyDerivedReferrer_MemberAddressListMap; }
    public String keepRegionId_SpecifyDerivedReferrer_MemberAddressList(MemberAddressCQ sq) {
        if (_regionId_SpecifyDerivedReferrer_MemberAddressListMap == null) { _regionId_SpecifyDerivedReferrer_MemberAddressListMap = newLinkedHashMapSized(4); }
        String ky = "subQueryMapKey" + (_regionId_SpecifyDerivedReferrer_MemberAddressListMap.size() + 1);
        _regionId_SpecifyDerivedReferrer_MemberAddressListMap.put(ky, sq); return "regionId_SpecifyDerivedReferrer_MemberAddressList." + ky;
    }

    protected Map<String, MemberAddressCQ> _regionId_InScopeRelation_MemberAddressListMap;
    public Map<String, MemberAddressCQ> getRegionId_InScopeRelation_MemberAddressList() { return _regionId_InScopeRelation_MemberAddressListMap; }
    public String keepRegionId_InScopeRelation_MemberAddressList(MemberAddressCQ sq) {
        if (_regionId_InScopeRelation_MemberAddressListMap == null) { _regionId_InScopeRelation_MemberAddressListMap = newLinkedHashMapSized(4); }
        String ky = "subQueryMapKey" + (_regionId_InScopeRelation_MemberAddressListMap.size() + 1);
        _regionId_InScopeRelation_MemberAddressListMap.put(ky, sq); return "regionId_InScopeRelation_MemberAddressList." + ky;
    }

    protected Map<String, MemberAddressCQ> _regionId_NotInScopeRelation_MemberAddressListMap;
    public Map<String, MemberAddressCQ> getRegionId_NotInScopeRelation_MemberAddressList() { return _regionId_NotInScopeRelation_MemberAddressListMap; }
    public String keepRegionId_NotInScopeRelation_MemberAddressList(MemberAddressCQ sq) {
        if (_regionId_NotInScopeRelation_MemberAddressListMap == null) { _regionId_NotInScopeRelation_MemberAddressListMap = newLinkedHashMapSized(4); }
        String ky = "subQueryMapKey" + (_regionId_NotInScopeRelation_MemberAddressListMap.size() + 1);
        _regionId_NotInScopeRelation_MemberAddressListMap.put(ky, sq); return "regionId_NotInScopeRelation_MemberAddressList." + ky;
    }

    protected Map<String, MemberAddressCQ> _regionId_QueryDerivedReferrer_MemberAddressListMap;
    public Map<String, MemberAddressCQ> getRegionId_QueryDerivedReferrer_MemberAddressList() { return _regionId_QueryDerivedReferrer_MemberAddressListMap; }
    public String keepRegionId_QueryDerivedReferrer_MemberAddressList(MemberAddressCQ sq) {
        if (_regionId_QueryDerivedReferrer_MemberAddressListMap == null) { _regionId_QueryDerivedReferrer_MemberAddressListMap = newLinkedHashMapSized(4); }
        String ky = "subQueryMapKey" + (_regionId_QueryDerivedReferrer_MemberAddressListMap.size() + 1);
        _regionId_QueryDerivedReferrer_MemberAddressListMap.put(ky, sq); return "regionId_QueryDerivedReferrer_MemberAddressList." + ky;
    }
    protected Map<String, Object> _regionId_QueryDerivedReferrer_MemberAddressListParameterMap;
    public Map<String, Object> getRegionId_QueryDerivedReferrer_MemberAddressListParameter() { return _regionId_QueryDerivedReferrer_MemberAddressListParameterMap; }
    public String keepRegionId_QueryDerivedReferrer_MemberAddressListParameter(Object vl) {
        if (_regionId_QueryDerivedReferrer_MemberAddressListParameterMap == null) { _regionId_QueryDerivedReferrer_MemberAddressListParameterMap = newLinkedHashMapSized(4); }
        String ky = "subQueryParameterKey" + (_regionId_QueryDerivedReferrer_MemberAddressListParameterMap.size() + 1);
        _regionId_QueryDerivedReferrer_MemberAddressListParameterMap.put(ky, vl); return "regionId_QueryDerivedReferrer_MemberAddressListParameter." + ky;
    }

    /** 
     * Add order-by as ascend. <br />
     * (地域ID)REGION_ID: {PK, NotNull, INTEGER(10), classification=Region}
     * @return this. (NotNull)
     */
    public BsRegionCQ addOrderBy_RegionId_Asc() { regOBA("REGION_ID"); return this; }

    /**
     * Add order-by as descend. <br />
     * (地域ID)REGION_ID: {PK, NotNull, INTEGER(10), classification=Region}
     * @return this. (NotNull)
     */
    public BsRegionCQ addOrderBy_RegionId_Desc() { regOBD("REGION_ID"); return this; }

    protected ConditionValue _regionName;
    public ConditionValue getRegionName() {
        if (_regionName == null) { _regionName = nCV(); }
        return _regionName;
    }
    protected ConditionValue getCValueRegionName() { return getRegionName(); }

    /** 
     * Add order-by as ascend. <br />
     * (地域名称)REGION_NAME: {NotNull, VARCHAR(50)}
     * @return this. (NotNull)
     */
    public BsRegionCQ addOrderBy_RegionName_Asc() { regOBA("REGION_NAME"); return this; }

    /**
     * Add order-by as descend. <br />
     * (地域名称)REGION_NAME: {NotNull, VARCHAR(50)}
     * @return this. (NotNull)
     */
    public BsRegionCQ addOrderBy_RegionName_Desc() { regOBD("REGION_NAME"); return this; }

    // ===================================================================================
    //                                                             SpecifiedDerivedOrderBy
    //                                                             =======================
    /**
     * Add order-by for specified derived column as ascend.
     * <pre>
     * cb.specify().derivedPurchaseList().max(new SubQuery&lt;PurchaseCB&gt;() {
     *     public void query(PurchaseCB subCB) {
     *         subCB.specify().columnPurchaseDatetime();
     *     }
     * }, <span style="color: #FD4747">aliasName</span>);
     * <span style="color: #3F7E5E">// order by [alias-name] asc</span>
     * cb.<span style="color: #FD4747">addSpecifiedDerivedOrderBy_Asc</span>(<span style="color: #FD4747">aliasName</span>);
     * </pre>
     * @param aliasName The alias name specified at (Specify)DerivedReferrer. (NotNull)
     * @return this. (NotNull)
     */
    public BsRegionCQ addSpecifiedDerivedOrderBy_Asc(String aliasName)
    { registerSpecifiedDerivedOrderBy_Asc(aliasName); return this; }

    /**
     * Add order-by for specified derived column as descend.
     * <pre>
     * cb.specify().derivedPurchaseList().max(new SubQuery&lt;PurchaseCB&gt;() {
     *     public void query(PurchaseCB subCB) {
     *         subCB.specify().columnPurchaseDatetime();
     *     }
     * }, <span style="color: #FD4747">aliasName</span>);
     * <span style="color: #3F7E5E">// order by [alias-name] desc</span>
     * cb.<span style="color: #FD4747">addSpecifiedDerivedOrderBy_Desc</span>(<span style="color: #FD4747">aliasName</span>);
     * </pre>
     * @param aliasName The alias name specified at (Specify)DerivedReferrer. (NotNull)
     * @return this. (NotNull)
     */
    public BsRegionCQ addSpecifiedDerivedOrderBy_Desc(String aliasName)
    { registerSpecifiedDerivedOrderBy_Desc(aliasName); return this; }

    // ===================================================================================
    //                                                                         Union Query
    //                                                                         ===========
    protected void reflectRelationOnUnionQuery(ConditionQuery bqs, ConditionQuery uqs) {
    }

    // ===================================================================================
    //                                                                       Foreign Query
    //                                                                       =============
    protected Map<String, Object> xfindFixedConditionDynamicParameterMap(String property) {
        return null;
    }

    // ===================================================================================
    //                                                                     ScalarCondition
    //                                                                     ===============
    protected Map<String, RegionCQ> _scalarConditionMap;
    public Map<String, RegionCQ> getScalarCondition() { return _scalarConditionMap; }
    public String keepScalarCondition(RegionCQ sq) {
        if (_scalarConditionMap == null) { _scalarConditionMap = newLinkedHashMapSized(4); }
        String ky = "subQueryMapKey" + (_scalarConditionMap.size() + 1);
        _scalarConditionMap.put(ky, sq); return "scalarCondition." + ky;
    }

    // ===================================================================================
    //                                                                       MyselfDerived
    //                                                                       =============
    protected Map<String, RegionCQ> _specifyMyselfDerivedMap;
    public Map<String, RegionCQ> getSpecifyMyselfDerived() { return _specifyMyselfDerivedMap; }
    public String keepSpecifyMyselfDerived(RegionCQ sq) {
        if (_specifyMyselfDerivedMap == null) { _specifyMyselfDerivedMap = newLinkedHashMapSized(4); }
        String ky = "subQueryMapKey" + (_specifyMyselfDerivedMap.size() + 1);
        _specifyMyselfDerivedMap.put(ky, sq); return "specifyMyselfDerived." + ky;
    }

    protected Map<String, RegionCQ> _queryMyselfDerivedMap;
    public Map<String, RegionCQ> getQueryMyselfDerived() { return _queryMyselfDerivedMap; }
    public String keepQueryMyselfDerived(RegionCQ sq) {
        if (_queryMyselfDerivedMap == null) { _queryMyselfDerivedMap = newLinkedHashMapSized(4); }
        String ky = "subQueryMapKey" + (_queryMyselfDerivedMap.size() + 1);
        _queryMyselfDerivedMap.put(ky, sq); return "queryMyselfDerived." + ky;
    }
    protected Map<String, Object> _qyeryMyselfDerivedParameterMap;
    public Map<String, Object> getQueryMyselfDerivedParameter() { return _qyeryMyselfDerivedParameterMap; }
    public String keepQueryMyselfDerivedParameter(Object vl) {
        if (_qyeryMyselfDerivedParameterMap == null) { _qyeryMyselfDerivedParameterMap = newLinkedHashMapSized(4); }
        String ky = "subQueryParameterKey" + (_qyeryMyselfDerivedParameterMap.size() + 1);
        _qyeryMyselfDerivedParameterMap.put(ky, vl); return "queryMyselfDerivedParameter." + ky;
    }

    // ===================================================================================
    //                                                                        MyselfExists
    //                                                                        ============
    protected Map<String, RegionCQ> _myselfExistsMap;
    public Map<String, RegionCQ> getMyselfExists() { return _myselfExistsMap; }
    public String keepMyselfExists(RegionCQ sq) {
        if (_myselfExistsMap == null) { _myselfExistsMap = newLinkedHashMapSized(4); }
        String ky = "subQueryMapKey" + (_myselfExistsMap.size() + 1);
        _myselfExistsMap.put(ky, sq); return "myselfExists." + ky;
    }

    // ===================================================================================
    //                                                                       MyselfInScope
    //                                                                       =============
    protected Map<String, RegionCQ> _myselfInScopeMap;
    public Map<String, RegionCQ> getMyselfInScope() { return _myselfInScopeMap; }
    public String keepMyselfInScope(RegionCQ sq) {
        if (_myselfInScopeMap == null) { _myselfInScopeMap = newLinkedHashMapSized(4); }
        String ky = "subQueryMapKey" + (_myselfInScopeMap.size() + 1);
        _myselfInScopeMap.put(ky, sq); return "myselfInScope." + ky;
    }

    // ===================================================================================
    //                                                                       Very Internal
    //                                                                       =============
    // very internal (for suppressing warn about 'Not Use Import')
    protected String xCB() { return RegionCB.class.getName(); }
    protected String xCQ() { return RegionCQ.class.getName(); }
    protected String xMap() { return Map.class.getName(); }
}

File: dbflute-scala-example/src/main/scala/com/example/dbflute/scala/dbflute/cbean/cq/bs/BsRegionCQ.scala
package com.example.dbflute.scala.dbflute.cbean.cq.bs;

import java.util.Map;

import org.seasar.dbflute.cbean._
import org.seasar.dbflute.cbean.chelper._
import org.seasar.dbflute.cbean.coption._
import org.seasar.dbflute.cbean.cvalue.ConditionValue;
import org.seasar.dbflute.cbean.sqlclause.SqlClause;
import org.seasar.dbflute.exception.IllegalConditionBeanOperationException;
import com.example.dbflute.scala.dbflute.cbean.cq.ciq._
import com.example.dbflute.scala.dbflute.cbean._
import com.example.dbflute.scala.dbflute.cbean.cq._

/**
 * The base condition-query of REGION.
 * @author DBFlute(AutoGenerator)
 */
class BsRegionCQ(referrerQuery: ConditionQuery, sqlClause: SqlClause, aliasName: String, nestLevel: Integer) extends AbstractBsRegionCQ(referrerQuery, sqlClause, aliasName, nestLevel) {

    // ===================================================================================
    //                                                                           Attribute
    //                                                                           =========
    protected var _inlineQuery: RegionCIQ = null;

    // ===================================================================================
    //                                                                 InlineView/OrClause
    //                                                                 ===================
    /**
     * Prepare InlineView query. <br />
     * {select ... from ... left outer join (select * from REGION) where FOO = [value] ...}
     * <pre>
     * cb.query().queryMemberStatus().<span style="color: #DD4747">inline()</span>.setFoo...;
     * </pre>
     * @return The condition-query for InlineView query. (NotNull)
     */
    def inline(): RegionCIQ = {
        if (_inlineQuery == null) { _inlineQuery = xcreateCIQ(); }
        _inlineQuery.xsetOnClause(false); return _inlineQuery;
    }

    protected def xcreateCIQ(): RegionCIQ = {
        val ciq: RegionCIQ = xnewCIQ();
        ciq.xsetBaseCB(_baseCB);
        return ciq;
    }

    protected def xnewCIQ(): RegionCIQ = {
        return new RegionCIQ(xgetReferrerQuery(), xgetSqlClause(), xgetAliasName(), xgetNestLevel(), this);
    }

    /**
     * Prepare OnClause query. <br />
     * {select ... from ... left outer join REGION on ... and FOO = [value] ...}
     * <pre>
     * cb.query().queryMemberStatus().<span style="color: #DD4747">on()</span>.setFoo...;
     * </pre>
     * @return The condition-query for OnClause query. (NotNull)
     * @throws IllegalConditionBeanOperationException When this condition-query is base query.
     */
    def on(): RegionCIQ = {
        if (isBaseQuery()) { throw new IllegalConditionBeanOperationException("OnClause for local table is unavailable!"); }
        val inlineQuery: RegionCIQ = inline(); inlineQuery.xsetOnClause(true); return inlineQuery;
    }

    // ===================================================================================
    //                                                                               Query
    //                                                                               =====
    protected var _regionId: ConditionValue = null;
    def getRegionId(): ConditionValue =
    { if (_regionId == null) { _regionId = nCV(); }
      return _regionId; }
    protected def getCValueRegionId(): ConditionValue = { return getRegionId(); }

    def getRegionId_ExistsReferrer_MemberAddressList(): Map[String, MemberAddressCQ] = { xgetSQueMap("regionId_ExistsReferrer_MemberAddressList") }
    def keepRegionId_ExistsReferrer_MemberAddressList(sq: MemberAddressCQ): String = { xkeepSQue("regionId_ExistsReferrer_MemberAddressList", sq) }

    def getRegionId_NotExistsReferrer_MemberAddressList(): Map[String, MemberAddressCQ] = { xgetSQueMap("regionId_NotExistsReferrer_MemberAddressList") }
    def keepRegionId_NotExistsReferrer_MemberAddressList(sq: MemberAddressCQ): String = { xkeepSQue("regionId_NotExistsReferrer_MemberAddressList", sq) }

    def getRegionId_SpecifyDerivedReferrer_MemberAddressList(): Map[String, MemberAddressCQ] = { xgetSQueMap("regionId_SpecifyDerivedReferrer_MemberAddressList") }
    def keepRegionId_SpecifyDerivedReferrer_MemberAddressList(sq: MemberAddressCQ): String = { xkeepSQue("regionId_SpecifyDerivedReferrer_MemberAddressList", sq) }

    def getRegionId_InScopeRelation_MemberAddressList(): Map[String, MemberAddressCQ] = { xgetSQueMap("regionId_InScopeRelation_MemberAddressList") }
    def keepRegionId_InScopeRelation_MemberAddressList(sq: MemberAddressCQ): String = { xkeepSQue("regionId_InScopeRelation_MemberAddressList", sq) }

    def getRegionId_NotInScopeRelation_MemberAddressList(): Map[String, MemberAddressCQ] = { xgetSQueMap("regionId_NotInScopeRelation_MemberAddressList") }
    def keepRegionId_NotInScopeRelation_MemberAddressList(sq: MemberAddressCQ): String = { xkeepSQue("regionId_NotInScopeRelation_MemberAddressList", sq) }

    def getRegionId_QueryDerivedReferrer_MemberAddressList(): Map[String, MemberAddressCQ] = { xgetSQueMap("regionId_QueryDerivedReferrer_MemberAddressList") }
    def keepRegionId_QueryDerivedReferrer_MemberAddressList(sq: MemberAddressCQ): String = { xkeepSQue("regionId_QueryDerivedReferrer_MemberAddressList", sq) }
    def getRegionId_QueryDerivedReferrer_MemberAddressListParameter(): Map[String, Object] = { xgetSQuePmMap("regionId_QueryDerivedReferrer_MemberAddressList") }
    def keepRegionId_QueryDerivedReferrer_MemberAddressListParameter(pm: Object): String = { xkeepSQuePm("regionId_QueryDerivedReferrer_MemberAddressList", pm) }

    /** 
     * Add order-by as ascend. <br />
     * (地域ID)REGION_ID: {PK, NotNull, INTEGER(10), classification=Region}
     * @return this. (NotNull)
     */
    def addOrderBy_RegionId_Asc(): BsRegionCQ = { regOBA("REGION_ID"); return this; }

    /**
     * Add order-by as descend. <br />
     * (地域ID)REGION_ID: {PK, NotNull, INTEGER(10), classification=Region}
     * @return this. (NotNull)
     */
    def addOrderBy_RegionId_Desc(): BsRegionCQ = { regOBD("REGION_ID"); return this; }

    protected var _regionName: ConditionValue = null;
    def getRegionName(): ConditionValue =
    { if (_regionName == null) { _regionName = nCV(); }
      return _regionName; }
    protected def getCValueRegionName(): ConditionValue = { return getRegionName(); }

    /** 
     * Add order-by as ascend. <br />
     * (地域名称)REGION_NAME: {NotNull, VARCHAR(50)}
     * @return this. (NotNull)
     */
    def addOrderBy_RegionName_Asc(): BsRegionCQ = { regOBA("REGION_NAME"); return this; }

    /**
     * Add order-by as descend. <br />
     * (地域名称)REGION_NAME: {NotNull, VARCHAR(50)}
     * @return this. (NotNull)
     */
    def addOrderBy_RegionName_Desc(): BsRegionCQ = { regOBD("REGION_NAME"); return this; }

    // ===================================================================================
    //                                                             SpecifiedDerivedOrderBy
    //                                                             =======================
    /**
     * Add order-by for specified derived column as ascend.
     * <pre>
     * cb.specify().derivedPurchaseList().max(new SubQuery&lt;PurchaseCB&gt;() {
     *     public void query(PurchaseCB subCB) {
     *         subCB.specify().columnPurchaseDatetime();
     *     }
     * }, <span style="color: #DD4747">aliasName</span>);
     * <span style="color: #3F7E5E">// order by [alias-name] asc</span>
     * cb.<span style="color: #DD4747">addSpecifiedDerivedOrderBy_Asc</span>(<span style="color: #DD4747">aliasName</span>);
     * </pre>
     * @param aliasName The alias name specified at (Specify)DerivedReferrer. (NotNull)
     * @return this. (NotNull)
     */
    def addSpecifiedDerivedOrderBy_Asc(aliasName: String): BsRegionCQ = { registerSpecifiedDerivedOrderBy_Asc(aliasName); return this; }

    /**
     * Add order-by for specified derived column as descend.
     * <pre>
     * cb.specify().derivedPurchaseList().max(new SubQuery&lt;PurchaseCB&gt;() {
     *     public void query(PurchaseCB subCB) {
     *         subCB.specify().columnPurchaseDatetime();
     *     }
     * }, <span style="color: #DD4747">aliasName</span>);
     * <span style="color: #3F7E5E">// order by [alias-name] desc</span>
     * cb.<span style="color: #DD4747">addSpecifiedDerivedOrderBy_Desc</span>(<span style="color: #DD4747">aliasName</span>);
     * </pre>
     * @param aliasName The alias name specified at (Specify)DerivedReferrer. (NotNull)
     * @return this. (NotNull)
     */
    def addSpecifiedDerivedOrderBy_Desc(aliasName: String): BsRegionCQ = { registerSpecifiedDerivedOrderBy_Desc(aliasName); return this; }

    // ===================================================================================
    //                                                                         Union Query
    //                                                                         ===========
    def reflectRelationOnUnionQuery(bqs: ConditionQuery, uqs: ConditionQuery): Unit = {
    }

    // ===================================================================================
    //                                                                       Foreign Query
    //                                                                       =============
    protected def xfindFixedConditionDynamicParameterMap(property: String): Map[String, Object] = {
        return null;
    }

    // ===================================================================================
    //                                                                     ScalarCondition
    //                                                                     ===============
    def getScalarCondition(): Map[String, RegionCQ] = { xgetSQueMap("scalarCondition") }
    def keepScalarCondition(sq: RegionCQ): String = { xkeepSQue("scalarCondition", sq) }

    // ===================================================================================
    //                                                                       MyselfDerived
    //                                                                       =============
    def getSpecifyMyselfDerived(): Map[String, RegionCQ] = { xgetSQueMap("specifyMyselfDerived") }
    def keepSpecifyMyselfDerived(sq: RegionCQ): String = { xkeepSQue("specifyMyselfDerived", sq) }

    def getQueryMyselfDerived(): Map[String, RegionCQ] = { xgetSQueMap("queryMyselfDerived") }
    def keepQueryMyselfDerived(sq: RegionCQ): String = { xkeepSQue("queryMyselfDerived", sq) }
    def getQueryMyselfDerivedParameter(): Map[String, Object] = { xgetSQuePmMap("qyeryMyselfDerived") }
    def keepQueryMyselfDerivedParameter(pm: Object): String = { xkeepSQuePm("qyeryMyselfDerived", pm) }

    // ===================================================================================
    //                                                                        MyselfExists
    //                                                                        ============
    def getMyselfExists(): Map[String, RegionCQ] = { xgetSQueMap("myselfExists") }
    def keepMyselfExists(sq: RegionCQ): String = { xkeepSQue("myselfExists", sq) }

    // ===================================================================================
    //                                                                       MyselfInScope
    //                                                                       =============
    def getMyselfInScope(): Map[String, RegionCQ] = { xgetSQueMap("myselfInScope") }
    def keepMyselfInScope(sq: RegionCQ): String = { xkeepSQue("myselfInScope", sq) }

    // ===================================================================================
    //                                                                       Very Internal
    //                                                                       =============
    // very internal (for suppressing warn about 'Not Use Import')
    protected def xCB(): String = { return classOf[RegionCB].getName(); }
    protected def xCQ(): String = { return classOf[RegionCQ].getName(); }
    protected def xCHp(): String = { return classOf[HpCalculator].getName(); }
    protected def xCOp(): String = { return classOf[ConditionOption].getName(); }
    protected def xMap(): String = { return classOf[Map[_, _]].getName(); }
}
